File: core/markdown_parser.py
"""
Knowledge Card の Markdown 形式を扱うモジュール。

仕様（仕様書 6章）:
    # タイトル

    ## Section名
    本文

    ## Section名
    本文

制約:
    - 同一Knowledge Card内でSection名は一意
    - Section順序はMarkdownの記載順
    - テンプレート固定は禁止、Sectionは自由追加可能

このモジュールは Markdown を「唯一の正データ」として扱うための
読み書きロジックのみを担当する。SQLite には一切依存しない。
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class Section:
    name: str
    content: str = ""


@dataclass
class KnowledgeDocument:
    title: str
    sections: List[Section] = field(default_factory=list)

    def section_names(self) -> List[str]:
        return [s.name for s in self.sections]

    def has_duplicate_section_names(self) -> bool:
        names = self.section_names()
        return len(names) != len(set(names))

    def get_section(self, name: str) -> Optional[Section]:
        for s in self.sections:
            if s.name == name:
                return s
        return None


_H1_RE = re.compile(r"^#\s+(.*?)\s*$")
_H2_RE = re.compile(r"^##\s+(.*?)\s*$")


def parse_markdown(text: str) -> KnowledgeDocument:
    """Markdown文字列を KnowledgeDocument にパースする。

    - 最初に見つかった "# " 見出しをタイトルとする
    - それ以降の "## " 見出しを Section の区切りとする
    - Section本文は次の "## " が出るまでの行を結合したもの
    """
    lines = text.replace("\r\n", "\n").split("\n")

    title = ""
    sections: List[Section] = []
    current_name: Optional[str] = None
    current_lines: List[str] = []
    title_found = False

    def flush():
        nonlocal current_name, current_lines
        if current_name is not None:
            content = "\n".join(current_lines).strip("\n")
            sections.append(Section(name=current_name, content=content))
        current_name = None
        current_lines = []

    for line in lines:
        h1_match = _H1_RE.match(line)
        h2_match = _H2_RE.match(line)

        if h1_match and not title_found:
            title = h1_match.group(1)
            title_found = True
            continue

        if h2_match:
            flush()
            current_name = h2_match.group(1)
            continue

        if current_name is not None:
            current_lines.append(line)
        # h1見出し発見前 & section開始前の行は無視（空行など）

    flush()
    return KnowledgeDocument(title=title, sections=sections)


def render_markdown(doc: KnowledgeDocument) -> str:
    """KnowledgeDocument を Markdown 文字列に書き出す。"""
    parts = [f"# {doc.title}\n"]
    for sec in doc.sections:
        content = sec.content.strip("\n")
        parts.append(f"\n## {sec.name}\n\n{content}\n")
    text = "\n".join(parts).strip("\n") + "\n"
    return text


def extract_title_only(text: str) -> str:
    """先頭の '# ' 行のみを高速に取り出す（一覧表示の再構築などで使用）。"""
    for line in text.replace("\r\n", "\n").split("\n"):
        m = _H1_RE.match(line)
        if m:
            return m.group(1)
    return ""

File: MedicalNoteDesktop/MedicalNoteDesktop/Core/MarkdownParser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace MedicalNoteDesktop.Core
{
    public class Section
    {
        public string Name { get; set; } = string.Empty;
        public string Content { get; set; } = string.Empty;
        public string UpdatedAt { get; set; } = string.Empty;
        public string UpdatedBy { get; set; } = string.Empty;
    }

    public class KnowledgeDocument
    {
        public string Title { get; set; } = string.Empty;
        public List<Section> Sections { get; set; } = new List<Section>();

        public Section? GetSection(string name)
        {
            return Sections.Find(s => s.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
        }
    }

    public static class MarkdownParser
    {
        private static readonly Regex UpdatedRegex = new Regex(@"<!--\s*updated:\s*(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}(?::\d{2})?)(?:\s*\((.*?)\))?\s*-->", RegexOptions.Compiled);
        private static readonly Regex OrganizedRegex = new Regex(@"<!--\s*organized:\s*(\d+)\s*-->\n?", RegexOptions.Compiled);

        public static KnowledgeDocument ParseMarkdown(string text)
        {
            var doc = new KnowledgeDocument();
            if (string.IsNullOrWhiteSpace(text)) return doc;

            string[] lines = text.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
            Section? currentSec = null;
            var currentLines = new List<string>();

            foreach (var line in lines)
            {
                if (line.StartsWith("# "))
                {
                    if (string.IsNullOrEmpty(doc.Title))
                    {
                        doc.Title = line.Substring(2).Trim();
                    }
                }
                else if (line.StartsWith("## "))
                {
                    if (currentSec != null)
                    {
                        currentSec.Content = string.Join("\n", currentLines).Trim();
                        doc.Sections.Add(currentSec);
                        currentLines.Clear();
                    }

                    currentSec = new Section
                    {
                        Name = line.Substring(3).Trim()
                    };
                }
                else
                {
                    if (currentSec != null)
                    {
                        // updated コメントのチェック
                        var match = UpdatedRegex.Match(line);
                        if (match.Success)
                        {
                            currentSec.UpdatedAt = match.Groups[1].Value;
                            currentSec.UpdatedBy = match.Groups[2].Value;
                        }
                        else
                        {
                            currentLines.Add(line);
                        }
                    }
                }
            }

            if (currentSec != null)
            {
                currentSec.Content = string.Join("\n", currentLines).Trim();
                doc.Sections.Add(currentSec);
            }

            return doc;
        }

        public static string RenderMarkdown(KnowledgeDocument doc)
        {
            var sb = new StringBuilder();
            sb.AppendLine($"# {doc.Title}");
            sb.AppendLine();

            foreach (var sec in doc.Sections)
            {
                sb.AppendLine($"## {sec.Name}");
                if (!string.IsNullOrEmpty(sec.UpdatedAt))
                {
                    string by = !string.IsNullOrEmpty(sec.UpdatedBy) ? $" ({sec.UpdatedBy})" : "";
                    sb.AppendLine($"<!-- updated:{sec.UpdatedAt}{by} -->");
                }
                if (!string.IsNullOrWhiteSpace(sec.Content))
                {
                    sb.AppendLine(sec.Content.Trim());
                }
                sb.AppendLine();
            }

            return sb.ToString().TrimEnd();
        }

        public static (string cleanContent, bool isOrganized) ParseOrganizedTag(string rawText)
        {
            bool isOrganized = rawText.Contains("<!-- organized: 1 -->");
            string clean = OrganizedRegex.Replace(rawText, "");
            return (clean, isOrganized);
        }

        public static string InjectOrganizedTag(string cleanText, bool isOrganized)
        {
            string clean = OrganizedRegex.Replace(cleanText, "");
            int val = isOrganized ? 1 : 0;
            if (clean.Contains("作成日時:"))
            {
                var parts = clean.Split(new[] { "作成日時:" }, 2, StringSplitOptions.None);
                var subparts = parts[1].Split(new[] { "\n" }, 2, StringSplitOptions.None);
                string dateLine = "作成日時:" + subparts[0];
                string rest = subparts.Length > 1 ? subparts[1] : "";
                return $"{parts[0]}{dateLine}\n<!-- organized: {val} -->\n{rest}";
            }
            return $"<!-- organized: {val} -->\n{clean}";
        }
    }
}
